src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -306,104 +357,526 @@ std::unique_ptr<Object::DeleteOp> D4NFilterObject::get_delete_op()
+
+        auto r = flush(dpp, std::move(completed));
+
+        if (r < 0) {
+          drain(dpp);
+          ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): Error: failed to flush, r= " << r << dendl;
+          return r;
+        }
+
+      } else {
+        ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): Info: draining data for oid: " << oid_in_cache << dendl;
+
+        auto r = drain(dpp);
+
+        if (r < 0) {
+          ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): Error: failed to drain, r= " << r << dendl;
+          return r;
+        }
+
+        break;
+      }
+    }
+
+    if (start_part_num == (num_parts - 1)) {
+      ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): Info: draining data for oid: " << oid_in_cache << dendl;
+      return drain(dpp);
+    } else {
+      adjusted_start_ofs += obj_max_req_size;
+    }
+
+    start_part_num += 1;
+    len -= obj_max_req_size;
+  } while (start_part_num < num_parts);
+
+  ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): Fetching object from backend store" << dendl;
+
+  Attrs obj_attrs;
+  if (source->has_attrs()) {
+    obj_attrs = source->get_attrs();
+  }
+
+  if (source->is_compressed() || obj_attrs.find(RGW_ATTR_CRYPT_MODE) != obj_attrs.end() || !y) {
+    ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): Skipping writing to cache" << dendl;
+    this->cb->bypass_cache_write();
+  }
+
+  if (start_part_num != 0) {
+    ofs = adjusted_start_ofs;
+  }
+
+  this->cb->set_ofs(ofs);
+  auto r = next->iterate(dpp, ofs, end, this->cb.get(), y);
+  
+  if (r < 0) {
+    ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): Error: failed to fetch object from backend store, r= " << r << dendl;
+    return r;
+  }
+
+  return this->cb->flush_last_part();
 }
 
-int D4NFilterWriter::prepare(optional_yield y) 
+int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::flush_last_part()
 {
-  int delDataReturn = filter->get_d4n_cache()->deleteData(obj->get_key().get_oid()); 
+  last_part = true;
+  return handle_data(bl_rem, 0, bl_rem.length());
+}
 
-  if (delDataReturn < 0) {
-    ldpp_dout(save_dpp, 20) << "D4N Filter: Cache delete data operation failed." << dendl;
-  } else {
-    ldpp_dout(save_dpp, 20) << "D4N Filter: Cache delete data operation succeeded." << dendl;
+int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data(bufferlist& bl, off_t bl_ofs, off_t bl_len)
+{
+  auto rgw_get_obj_max_req_size = g_conf()->rgw_get_obj_max_req_size;
+
+  if (!last_part && bl.length() <= rgw_get_obj_max_req_size) {
+    auto r = client_cb->handle_data(bl, bl_ofs, bl_len);
+
+    if (r < 0) {
+      return r;
+    }
   }
 
+  //Accumulating data from backend store into rgw_get_obj_max_req_size sized chunks and then writing to cache
+  if (write_to_cache) {
+    const std::lock_guard l(d4n_get_data_lock);
```

Comment:
this lock has a huge scope; avoid holding mutexes over i/o like get_cache_driver()->put_async() and get_cache_policy()->update()

Reply:
Yes, checking all the locks is also on my TODO list

Reply:
within a single `D4NFilterReadOp`, i don't think you need that mutex at all

the AioThrottle from `aio = rgw::make_throttle()` should handle any thread-safety issues for you (where io completions happen in a different background thread), as long as you're only accessing `aio` in `D4NFilterReadOp::iterate()`

Reply:
the lock here is for writing to the cache backend - wouldn't that be needed if we have two parallel `get` calls, and both of them read from the backend store and then try to write to the backend? Can such a scenario not arise?